src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.Encrypt.cs

```diff
@@ -207,36 +207,21 @@ public static SafeCryptMsgHandle CreateCryptMsgHandleToEncode(CmsRecipientCollec
             //
             private static CMSG_RECIPIENT_ENCODE_INFO EncodeRecipientInfo(CmsRecipient recipient, AlgorithmIdentifier contentEncryptionAlgorithm, HeapBlockRetainer hb)
             {
-                CMsgCmsRecipientChoice recipientChoice;
-                string keyAlgorithmOid = recipient.Certificate.GetKeyAlgorithm();
-                AlgId algId = keyAlgorithmOid.ToAlgId();
-                switch (algId)
-                {
-                    case AlgId.CALG_RSA_KEYX:
-                        recipientChoice = CMsgCmsRecipientChoice.CMSG_KEY_TRANS_RECIPIENT;
-                        break;
-
-                    case AlgId.CALG_DH_SF:
-                    case AlgId.CALG_DH_EPHEM:
-                        recipientChoice = CMsgCmsRecipientChoice.CMSG_KEY_AGREE_RECIPIENT;
-                        break;
-
-                    default:
-                        throw ErrorCode.CRYPT_E_UNKNOWN_ALGO.ToCryptographicException();
-                }
-
                 CMSG_RECIPIENT_ENCODE_INFO recipientEncodeInfo;
-                recipientEncodeInfo.dwRecipientChoice = recipientChoice;
 
                 unsafe
                 {
-                    switch (recipientChoice)
+                    switch (recipient.Certificate.GetKeyAlgorithm())
                     {
-                        case CMsgCmsRecipientChoice.CMSG_KEY_TRANS_RECIPIENT:
+                        case Oids.Rsa:
+                        case Oids.RsaOaep:
+                            recipientEncodeInfo.dwRecipientChoice = CMsgCmsRecipientChoice.CMSG_KEY_TRANS_RECIPIENT;
                             recipientEncodeInfo.pCmsRecipientEncodeInfo = (IntPtr)EncodeKeyTransRecipientInfo(recipient, hb);
                             break;
 
-                        case CMsgCmsRecipientChoice.CMSG_KEY_AGREE_RECIPIENT:
+                        case Oids.Esdh:
```

Comment:
To my knowledge these are the only two OIDs that map to `CMSG_KEY_AGREE_RECIPIENT`. Is there a known map anywhere?

Reply:
Since the previous thing was using CALG_ID values, it looks like Windows does

* CALG_DH_SF
  * 1.2.840.10046.2.1
  * 1.2.840.113549.1.3.1
* CALG_DH_EPHEM
  * 1.2.840.113549.1.9.16.3.5

So, for maximum compatibility, I guess add the missing one as Oids.DiffieHellmanPkcs3, or something like that?  (FWIW https://tools.ietf.org/html/rfc3279#section-2.3.3 says to use 1.2.840.10046.2.1)